src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,505 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/BasicConfig.h>
+#include <ripple/basics/impl/PerfLogImp.h>
+#include <ripple/beast/core/CurrentThreadName.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/json/json_writer.h>
+#include <ripple/json/to_string.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <atomic>
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+namespace ripple {
+namespace perf {
+
+void
+PerfLogImp::Counters::populateRpc(std::vector<char const*> const& labels)
+{
+    for ( auto label : labels)
+    {
+        auto const &elem = rpc_.emplace(label, Rpc(label));
+        if (!elem.second)
+            assert(false);
```

Comment:
There's nothing in this PR that should cause a production server to crash. I don't want to turn a diagnostics collection bug into a crashing bug.

Reply:
But an `assert` is a crash of sorts.  And an `assert` is what you've included in your code.  The `assert` is a way of saying "this should never happen".  What I'm suggesting is only a little bit stronger.  By using `LogicError` instead of `assert` you'd be saying "this should never happen, even in a release build".  And since both of these operations happen early in the boot phase of the server you'd be shutting down a server that was 1) not yet fully operational and 2) had a logic error.

But I don't feel strongly about it.  If you feel an `assert` is a better choice I'm okay with that.

Reply:
Yes, I think assert is a better choice in this case to bias towards runtime stability vs diagnostics. For transaction processing or something delivered to users, it's better to crash immediately, but that's not this.